samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -19,258 +19,212 @@
-    var msgCount = 0
     while (iterator.hasNext) {
-      val msg = iterator.next
-      msgCount += 1
-
-      val offset = msg.getOffset
-      val key = msg.getKey.asInstanceOf[Array[Byte]]
-      if (key == null) {
-        throw new KafkaUtilException(
-          "While reading checkpoint (currentOffset=%s) stream encountered message without key." format offset)
-      }
+      numMessagesRead += 1
+      val checkpointEnvelope: IncomingMessageEnvelope = iterator.next
 
-      val checkpointKey = KafkaCheckpointLogKey.fromBytes(key)
+      val keyBytes = checkpointEnvelope.getKey.asInstanceOf[Array[Byte]]
+      if (keyBytes == null) {
+        throw new SamzaException(s"Encountered a checkpoint message without key. Topic:$checkpointTopic " +
+          s"Offset:${checkpointEnvelope.getOffset}")
+      }
 
-      if (!shouldHandleEntry(checkpointKey)) {
-        info("Skipping checkpoint log entry at offset %s with key %s." format(offset, checkpointKey))
-      } else {
-        val checkpointPayload = ByteBuffer.wrap(msg.getMessage.asInstanceOf[Array[Byte]])
-        handleEntry(checkpointPayload, checkpointKey)
+      val checkpointKey = checkpointKeySerde.fromBytes(keyBytes)
+      val expectedGrouperFactory = checkpointKey.getGrouperFactoryClassName
+      if (!grouperFactory.equals(expectedGrouperFactory)) {
+        throw new SamzaException("SSPGrouperFactory in the checkpoint topic does not match the configured value" +
+          s"Found: $grouperFactory Expected: $expectedGrouperFactory")
       }
-    }
-    info("Done reading %s messages from checkpoint system:%s topic:%s" format(msgCount, systemName, checkpointTopic))
-  }
 
-  override def start {
-    val CHECKPOINT_STREAMID = "unused-temp-checkpoint-stream-id"
-    val spec = new KafkaStreamSpec(CHECKPOINT_STREAMID,
-                                   checkpointTopic, systemName, 1,
-                                   replicationFactor, checkpointTopicProperties)
-
-    info("About to create checkpoint stream: " + spec)
-    systemAdmin.createStream(spec)
-    info("Created checkpoint stream: " + spec)
-    try {
-      systemAdmin.validateStream(spec) // SPECIAL VALIDATION FOR CHECKPOINT. DO NOT FAIL IF failOnCheckpointValidation IS FALSE
-      info("Validated spec: " + spec)
-    } catch {
-      case e : StreamValidationException =>
-             if (failOnCheckpointValidation) {
-               throw e
-             } else {
-               warn("Checkpoint stream validation partially failed. Ignoring it because failOnCheckpointValidation=" + failOnCheckpointValidation)
-             }
-      case e1 : Exception => throw e1
+      val checkpointBytes = checkpointEnvelope.getMessage.asInstanceOf[Array[Byte]]
+      val checkpoint = checkpointMsgSerde.fromBytes(checkpointBytes)
+      checkpoints.put(checkpointKey.getTaskName, checkpoint)
     }
+    info(s"Read $numMessagesRead messages from system:$checkpointSystem topic:$checkpointTopic")
```

Comment:
>> I think it'll be very useful if we log progress in the loop

Great suggestion, added a progress marker! Also logged the offset just in case we want to track. 

>> We can also log a warning if numMessagesRead happens to be much larger (for some multiplier) than num tasks. This would imply that the topic isn't compacted.

I'm not sure if the warning / heuristic multiplier is particularly useful. We'll have to look at logs anyways to debug these issues.